File: SUAPC/2024W/H.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(5*int(1e5)+1000)

N = int(input())
lst = [input().rstrip() for i in range(N)]
nxt = [i for i in range(N+1)]
front = [i for i in range(N+1)]
back = [i for i in range(N+1)]

def findfront(a):
    if a == front[a]:
        return a
    front[a] = findfront(front[a])
    return front[a]

def findback(a):
    if a == back[a]:
        return a
    back[a] = findback(back[a])
    return back[a]

def merge(a, b):
    af = findfront(a)
    ab = findback(a)
    bf = findfront(b)
    bb = findback(b)
    front[bf] = af
    back[ab] = bb
    nxt[ab] = bf

for i in range(N-1):
    a, b = map(int, input().split())
    merge(a, b)

idx = findfront(1)
while idx != nxt[idx]:
    print(lst[idx-1], end="")
    idx = nxt[idx]
print(lst[idx-1])

File: SUAPC/2024W/H.cpp
#include <bits/stdc++.h>
using namespace std;

int N;
vector<string> lst;
vector<int> nxt, front, back;

int findfront(int a) {
    if (a == front[a]) return a;
    front[a] = findfront(front[a]);
    return front[a];
}

int findback(int a) {
    if (a == back[a]) return a;
    back[a] = findback(back[a]);
    return back[a];
}

void merge(int a, int b) {
    int af = findfront(a);
    int ab = findback(a);
    int bf = findfront(b);
    int bb = findback(b);
    front[bf] = af;
    back[ab] = bb;
    nxt[ab] = bf;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> N;
    lst.resize(N);
    nxt.resize(N + 1);
    front.resize(N + 1);
    back.resize(N + 1);

    for (int i = 1; i <= N; ++i) {
        cin >> lst[i - 1];
        nxt[i] = i;
        front[i] = i;
        back[i] = i;
    }

    for (int i = 0; i < N - 1; ++i) {
        int a, b;
        cin >> a >> b;
        merge(a, b);
    }

    int idx = findfront(1);
    while (idx != nxt[idx]) {
        cout << lst[idx - 1];
        idx = nxt[idx];
    }
    cout << lst[idx - 1];

    return 0;
}
